### ISSUE TITLE
Crash when probing without setting data to a buffer

### ISSUE THREAD

**Issue by cmarcelo** [google/amber#501]
After d26ee22 "Unify buffer sizes. (#465)", using probe in a buffer without previously setting it will cause crash apparently due to format not being set. Seems to me that is a bug since would be fine for a buffer to be used as output and only probed at the end.

- If that's the case then probe should also be able to "set the format" of a buffer.

- If that's not the case: would be nice if this behavior was documented and, since amber can identify those cases, an proper error should be printed.

Snippet of the stack trace:

```
#0  amber::Format::SizeInBytes (this=0x0)
#1  0x0000555556b91ced in amber::vulkan::BufferDescriptor::MoveResourceToBufferOutput (this=0x555568758720) at /usr/include/c++/9/bits/unique_ptr.h:357
#2  0x0000555556b8e732 in amber::vulkan::Pipeline::ReadbackDescriptorsToHostDataQueue (this=this@entry=0x55556571b730) at /usr/include/c++/9/bits/unique_ptr.h:357
#3  0x0000555556b9293a in amber::vulkan::ComputePipeline::Compute (this=0x55556571b730, x=<optimized out>, y=1, z=1)
#4  0x0000555556b892f4 in amber::vulkan::EngineVulkan::DoCompute (this=this@entry=0x555568757ac0, command=0x5555687269e0)
#5  0x0000555556b5188a in amber::Executor::ExecuteCommand (this=this@entry=0x7fffffffb9cf, engine=engine@entry=0x555568757ac0, cmd=0x5555687269e0)
#6  0x0000555556b51ac9 in amber::Executor::Execute (this=this@entry=0x7fffffffb9cf, engine=0x555568757ac0, script=<optimized out>, shader_map=..., executionType=amber::ExecutionType::kExecute)
    at /usr/include/c++/9/bits/unique_ptr.h:357
#7  0x0000555556b42342 in amber::Amber::ExecuteWithShaderData (this=<optimized out>, recipe=<optimized out>, opts=0x7fffffffbc40, shader_data=...) at /usr/include/c++/9/bits/unique_ptr.h:357
```

**Comment by cmarcelo** [comment]
Example snippet that leads to a crash. Uncommenting the subdatas there's no crash.

```
[test]
ssbo 0:0 2048
ssbo 0:1 2176
uniform ubo 0:2 int 0 32

# ssbo 0:0 subdata int 0  0
# ssbo 0:1 subdata int 0  0

compute entrypoint main
compute 32 1 1

probe ssbo int 0:1 0    == 0 1 1 1 1 0 1 1 1 1 0 1 1 1 1 0 0
```

**Comment by dj2** [comment]
This isn't the probe which is crashing. When we execute the `compute` command we attempt to read the data back from the Device to the Host. That readback needs the format in order to be able to read the data back.

We should be detecting that the buffers aren't complete before starting to execute any commands and exit with an error.

We could possibly use the probe command to set the format on the buffer, will have to think about it to see if that would be confusing.

AmberScript does not have this issue as the buffer always has a format attached when it's created.

**Comment by dj2** [comment]
Actually, it already does that, hm. If the probe finds a buffer without a format the format is set. Need to dig a bit more to see what's happening.

**Comment by dj2** [comment]
Ah, the problem is the ssbo at 0:0 which has no format. The one at 0:1 gets the format from the probe, but 0:0 has no format and we fail to read back.

**Comment by dj2** [comment]
So, removing ssbo 0:0 2048 in the above example makes the test run as expected.

**Comment by cmarcelo** [comment]
Some tests may use a buffer for scratch area. Would make sense to not readback/validate from buffers that we don't need to probe?

**Comment by dj2** [comment]
Maybe? It makes the code more convoluted as you have to know what buffers are probed. We've currently gone with the simplest route of always read back. If you want a scratch buffer, you can always just poke a single value in there with the subdata command to set the type. It's a little cumbersome, but saves a lot of magic on the backend.

**Comment by cmarcelo** [comment]
> If you want a scratch buffer, you can always just poke a single value in there with the subdata command to set the type. It's a little cumbersome, but saves a lot of magic on the backend.

The scratch area usage currently causes a crash, it probably should print an error message instead then?

Also would be nice to have a note about this restriction in the vk_script documentation.

```
[compute shader spirv]
OpCapability Shader
OpMemoryModel Logical GLSL450
OpEntryPoint GLCompute %main "main"
OpExecutionMode %main LocalSize 1 1 1

%void = OpTypeVoid

%void_fn = OpTypeFunction %void
%main = OpFunction %void None %void_fn
%entry = OpLabel

OpReturn
OpFunctionEnd

[test]
ssbo 0:1 8

compute entrypoint main
compute 1 1 1
```

**Comment by dj2** [comment]
Printing an error is what PR #502 does. Is an error correctly printed with that CL?

**Comment by dj2** [comment]
Reopening until the documentation change lands.

**Comment by dneto0** [comment]
I just worked around this issue in a few internal tests by adding the subdata.  It's a sharp edge on the tool that we should strive to fix.

**Comment by dj2** [comment]
PR #514 should allow the usage of scratch buffers.

**Comment by cmarcelo** [comment]
> PR #514 should allow the usage of scratch buffers.

Thanks, Dan.
